File: py/src/aoc2020/day14.py
from dataclasses import dataclass
import fileinput
from functools import reduce
from itertools import chain, product


@dataclass
class Mask:
    off: int
    on: int


@dataclass
class Write:
    addr: int
    value: int


def parse(line):
    if line.startswith('mask = '):
        mask = line[7:]
        return Mask(off=int(mask.replace('X', '0'), base=2),
                    on=int(mask.replace('X', '1'), base=2))
    if line.startswith('mem['):
        i = line.index('] = ')
        return Write(addr=int(line[4:i]), value=int(line[i + 4:]))


def part1(lines):
    '''
    >>> part1(('mask = XXXXXXXXXXXXXXXXXXXXXXXXXXXXX1XXXX0X', 'mem[8] = 11', 'mem[7] = 101', 'mem[8] = 0'))
    165
    '''
    mem = {}
    for line in lines:
        instruction = parse(line)
        if isinstance(instruction, Mask):
            maskoff, maskon = instruction.off, instruction.on
        elif isinstance(instruction, Write):
            mem[instruction.addr] = (instruction.value | maskoff) & maskon
    return sum(mem.values())


def part2(lines):
    '''
    >>> part2(('mask = 000000000000000000000000000000X1001X', 'mem[42] = 100', 'mask = 00000000000000000000000000000000X0XX', 'mem[26] = 1'))
    208
    '''
    mem = {}
    for line in lines:
        instruction = parse(line)
        if isinstance(instruction, Mask):
            maskoff, maskon = instruction.off, instruction.on
        elif isinstance(instruction, Write):
            diff = maskoff ^ maskon
            popcount = bin(diff).count('1')
            for i in range(1 << popcount):

                def f(acc, j):
                    x, k = acc
                    return (x ^ diff & (k ^ k - (i >> j & 1)), k & k - 1)

                mem[reduce(
                    f, range(popcount),
                    (instruction.addr | maskoff, diff))[0]] = instruction.value
    return sum(mem.values())


parts = (part1, part2)

if __name__ == '__main__':
    lines = list(fileinput.input())
    print(part1(lines))
    print(part2(lines))

File: rs/src/day14.rs
use std::collections::HashMap;
use std::error::Error;
use std::fmt::{self, Display, Formatter};
use std::num::ParseIntError;
use std::str::FromStr;

#[derive(Clone, Debug)]
enum ParseInstructionErr {
    Format,
    Value(ParseIntError),
}

impl Display for ParseInstructionErr {
    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
        match self {
            Self::Format => write!(f, "ParseInstructionErr::Format"),
            Self::Value(err) => write!(f, "ParseInstructionErr::Value({})", err),
        }
    }
}

impl Error for ParseInstructionErr {}

enum Instruction {
    Mask { off: u64, on: u64 },
    Write { addr: u64, value: u64 },
}

impl FromStr for Instruction {
    type Err = ParseInstructionErr;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        if let Some(mask) = s.strip_prefix("mask = ") {
            Ok(Instruction::Mask {
                off: u64::from_str_radix(&mask.replace('X', "0"), 2)
                    .map_err(ParseInstructionErr::Value)?,
                on: u64::from_str_radix(&mask.replace('X', "1"), 2)
                    .map_err(ParseInstructionErr::Value)?,
            })
        } else if let Some((addr, value)) = s.strip_prefix("mem[").and_then(|s| {
            let mut iter = s.splitn(2, "] = ");
            Some((iter.next()?, iter.next()?))
        }) {
            Ok(Instruction::Write {
                addr: addr.parse().map_err(ParseInstructionErr::Value)?,
                value: value.parse().map_err(ParseInstructionErr::Value)?,
            })
        } else {
            Err(ParseInstructionErr::Format)
        }
    }
}

pub fn part1<'a, I, S>(lines: I) -> Result<u64, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let mut mem = HashMap::new();
    let mut mask = (0, 0);
    for line in lines.into_iter() {
        match line.as_ref().parse()? {
            Instruction::Mask { off, on } => mask = (off, on),
            Instruction::Write { addr, value } => {
                mem.insert(addr, (value | mask.0) & mask.1);
            }
        }
    }
    Ok(mem.values().sum())
}

pub fn part2<'a, I, S>(lines: I) -> Result<u64, Box<dyn Error + Send + Sync>>
where
    I: IntoIterator<Item = &'a S>,
    S: AsRef<str> + 'a,
{
    let mut mem = HashMap::new();
    let mut mask = (0, 0);
    for line in lines.into_iter() {
        match line.as_ref().parse()? {
            Instruction::Mask { off, on } => mask = (off, on),
            Instruction::Write { addr, value } => {
                let diff = mask.0 ^ mask.1;
                for i in 0..1 << diff.count_ones() {
                    mem.insert(
                        (0..diff.count_ones())
                            .fold((addr | mask.0, diff), |(addr2, diff2), j| {
                                (
                                    addr2 ^ diff & (diff2 ^ (diff2 - (i >> j & 1))),
                                    diff2 & (diff2 - 1),
                                )
                            })
                            .0,
                        value,
                    );
                }
            }
        }
    }
    Ok(mem.values().sum())
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn part1_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(
            165,
            part1(&[
                "mask = XXXXXXXXXXXXXXXXXXXXXXXXXXXXX1XXXX0X",
                "mem[8] = 11",
                "mem[7] = 101",
                "mem[8] = 0"
            ])?
        );
        Ok(())
    }

    #[test]
    fn part2_examples() -> Result<(), Box<dyn Error + Send + Sync>> {
        assert_eq!(
            208,
            part2(&[
                "mask = 000000000000000000000000000000X1001X",
                "mem[42] = 100",
                "mask = 00000000000000000000000000000000X0XX",
                "mem[26] = 1"
            ])?
        );
        Ok(())
    }
}
